sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,574 @@
+
+            // Take the anonymous type representing the binding and serialize it as a JSON string
+            SerializeAnonymousBindingAsJsonString(indentedTextWriter, functionName, bindingCount.ToString());
+            bindingCount++;
+
+            return bindingCount;
+        }
+
+        private static void SerializeAnonymousBindingAsJsonString(IndentedTextWriter indentedTextWriter, string functionName, string bindingCount)
+        {
+            indentedTextWriter.WriteLine($"var {functionName}Binding{bindingCount}JSONstring = JsonSerializer.Serialize({functionName}Binding{bindingCount});");
+            indentedTextWriter.WriteLine($"{functionName}RawBindings.Add({functionName}Binding" + bindingCount.ToString() + "JSONstring);");
+        }
+
+        private static void AddEnumTypes(IndentedTextWriter indentedTextWriter)
+        {
+            indentedTextWriter.WriteLine("public enum AuthorizationLevel");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("Anonymous,");
+            indentedTextWriter.WriteLine("User,");
+            indentedTextWriter.WriteLine("Function,");
+            indentedTextWriter.WriteLine("System,");
+            indentedTextWriter.WriteLine("Admin");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        /// <summary>
+        /// Adds a generated registration extension that users can call to register the source-generated function metadata provider.
+        /// </summary>
+        private static void AddRegistrationExtension(IndentedTextWriter indentedTextWriter)
+        {
+            indentedTextWriter.WriteLine("public static class WorkerHostBuilderFunctionMetadataProviderExtension");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("public static IHostBuilder ConfigureGeneratedFunctionMetadataProvider(this IHostBuilder builder)");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("builder.ConfigureServices(s => ");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("s.AddSingleton<IFunctionMetadataProvider, GeneratedFunctionMetadataProvider>();");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("});");
+            indentedTextWriter.WriteLine("return builder;");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        private static bool IsStringType(string fullName)
+        {
+            return string.Equals(fullName, "String", StringComparison.Ordinal);
+        }
+
+        // TODO: Test/Verify this
```

Comment:
In progress: writing an e2e test with a function using a binary type to validate this so ignore this method for now.